### PR TITLE
Refactor rounding (including bugfixes) and document API

### LINKED ISSUE #42060: Extend floor (round, etc) to Complex numbers
I was recently porting a Matlab example to julia and I realized there's no direct equivalence.

I extended the floor function in my code snippet by doing:
```julia
import Base.floor
floor(x::Complex) = floor(real(x)::Real) + 1im*floor(imag(x)::Real)
```

I'm not sure if this would be a welcome addition, but it would surely make Julia code more similar to other well-used scientific software.

### LINKED ISSUE #50778: `round(Float64, 1.2)` is documented to return `1.0`, but throws.
```julia

help?> round(Float64, 1.2)
  round([T,] x, [r::RoundingMode])
  round(x, [r::RoundingMode]; digits::Integer=0, base = 10)
  round(x, [r::RoundingMode]; sigdigits::Integer, base = 10)

  Rounds the number x.

  Without keyword arguments, x is rounded to an integer value, returning a value 
  of type T, or of the same type of x if no T is provided. An InexactError will be 
  thrown if the value is not representable by T, similar to convert.

...

julia> round(Float64, 1.2)
ERROR: MethodError: no method matching round(::Type{Float64}, ::Float64)

Closest candidates are:
  round(::Type{Bool}, ::AbstractFloat)
   @ Base float.jl:391
  round(::Type{T}, ::AbstractFloat, ::RoundingMode) where T<:Integer
   @ Base floatfuncs.jl:122
  round(::Type{T}, ::AbstractFloat) where T<:Integer
   @ Base float.jl:385
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[13]:1
```

### LINKED ISSUE #51113: `round(Int, -Inf16)` doesn't throw
```julia
julia> round(Int32, -Inf16)
-2147483648

julia> round(Int64, -Inf16)
-9223372036854775808

julia> round(Int128, -Inf16)
0

julia> versioninfo()
Julia Version 1.10.0-beta2
Commit a468aa198d0 (2023-08-17 06:27 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M2
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
  Threads: 1 on 4 virtual cores
Environment:
  JULIA_EDITOR = code
```

### PR DESCRIPTION
**Pull request by LilithHafner**
This PR implements and documents a cohesive, extensible rounding interface. The code that implements it is in base/rounding.jl

```julia
trunc(x; kwargs...) = round(x, RoundToZero; kwargs...)
floor(x; kwargs...) = round(x, RoundDown; kwargs...)
 ceil(x; kwargs...) = round(x, RoundUp; kwargs...)
round(x; kwargs...) = round(x, RoundNearest; kwargs...)

trunc(::Type{T}, x; kwargs...) where {T} = round(T, x, RoundToZero; kwargs...)
floor(::Type{T}, x; kwargs...) where {T} = round(T, x, RoundDown; kwargs...)
 ceil(::Type{T}, x; kwargs...) where {T} = round(T, x, RoundUp; kwargs...)
round(::Type{T}, x; kwargs...) where {T} = round(T, x, RoundNearest; kwargs...)

round(::Type{T}, x, r::RoundingMode) where {T} = convert(T, round(x, r))
```

And an explanation of how to use it is in docs/src/manual/interfaces.md.

Fixes #50778
Fixes #42060, though maybe this is a bad thing.
Closes #47128
Fixes #51113

This PR also refactors some implementations of rounding for concrete types to use this new interface more naturally.

### CODE REVIEW

**Comment by oscardssmith** [comment]
adding triage label to this to discuss whether Inf/-Inf are valid values for the return. Imo inf can't be a valid representation of a rounded number.

**Comment by LilithHafner** [comment]
I am inclined to allow inf for reasons I mentioned in https://github.com/JuliaLang/julia/issues/50778#issuecomment-1666899328 and concur that we should discuss this at triage.

**Comment by adienes** [comment]
without knowing too much of the details here, I think it could be nice to coordinate with https://github.com/JuliaLang/julia/pull/50730 such that

```
isapprox(x, round(T, x); atol=1)
```

is always true (or throws an error)

**Comment by LilithHafner** [comment]
That would be nice, though [hypercomplex numbers](https://en.wikipedia.org/wiki/Hypercomplex_number) can be arbitrarily far from the nearest lattice point.

```julia
julia> x = Complex(1.9, 1.9)
1.9 + 1.9im

julia> isapprox(x, round(x, RoundDown), atol=1)
false

julia> isapprox(x, round(x, RoundDown), atol=1.3)
true
```

With greater than 4 dimensions it is possible to get this behavior with `RoundNearest` as well.

```julia
julia> using Octanians

julia> x = Octonion(fill(1.5, 8)...)
OctonionF64(1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5)

julia> Base.round(x::Octonion, r::RoundingMode) = Octonion(
           round(x.s, r),
           round(x.v1, r),
           round(x.v2, r),
           round(x.v3, r),
           round(x.v4, r),
           round(x.v5, r),
           round(x.v6, r),
           round(x.v7, r))

julia> y = round(x)
OctonionF64(2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0)

julia> isapprox(x, y, atol=1)
false

julia> isapprox(x, y, atol=1.5)
true
```

**Comment by adienes** [comment]
well, maybe only guarantee that holds for `<:Real` then 😅, I guess there will always be niche number types for edge cases

**Comment by nsajko** [comment]
> `@test round(Float32, 1e60) == Inf32`

This is wrong because infinity isn't an integer. The Julia documentation for `round` already says that the return value is integral, so this conflicts with the documented behavior. Also note that `unsafe_trunc` is already available and part of the public interface.

The more interesting question is what should be done when the user provides an infinite value as input. FWIW, C/C++ return infinity (with unchanged sign): https://en.cppreference.com/w/cpp/numeric/math/round https://en.cppreference.com/w/cpp/numeric/math/rint https://en.cppreference.com/w/cpp/numeric/math/nearbyint

**Comment by Seelengrab** [comment]
Related https://github.com/JuliaLang/julia/issues/42060b

**Comment by vchuravy** [comment]
Could we split the bugfix out so that we can backport it?

**Comment by Seelengrab** [comment]
For the record, I don't think this PR should implement #42060, for the reasons given in the discussion there.

**Comment by LilithHafner** [comment]
> Could we split the bugfix out so that we can backport it?

The bugfix is implementing a missing method by widening type signatures and adding default fallbacks, so no, this would have to be backported manually.

**Comment by mikmoore** [comment]
Nice!

My obnoxious corner-case with
```
round(::Type{T}, x, r::RoundingMode) where {T} = convert(T, round(x, r))
```
is to complain that I would naively expect `round(Float32, 2^30+1, RoundUp)` to produce `nextfloat(2f0^30) == 2f0^30+128f0`, but with this chain of logic it would `round` `2^30+1` as a no-op, then `convert` this to `2f0^30`. But perhaps this really should be the domain of `convert` and not `round` at all, so perhaps this shouldn't be considered here...

**Comment by imciner2** [comment]
> adding triage label to this to discuss whether Inf/-Inf are valid values for the return. Imo inf can't be a valid representation of a rounded number.

Wouldn't this be dependent on the rounding mode and type being rounded to? For Float types, IEEE semantics for the rounding could dictate when/if `Inf`/`-Inf` is returned based on Sec 4.3 of the standard (and Sec 7.4 for the overflow operation for the inexact semantics). For instance, with these semantics it is impossible for `RoundToZero` to ever return an `Inf`/`-Inf` from the function, but `RoundUp` can return `Inf` only when the given value is greater than the largest representable number of the format (and never return a `-Inf`).

**Comment by nsajko** [comment]
@imciner2 there are two different meanings of round at play here. The parts of the standard you refer to are concerned with rounding to some value in the floating-point format, while this PR deals with rounding to integer values.

The standard also covers that topic, in 5.9, "Details of operations to round a floating-point datum to integral value". Excerpt:

> The rounding is analogous to that specified in Clause 4, but the rounding chooses only from among those
floating-point numbers of integral values in the format. These operations convert zero operands to zero
results of the same sign, and infinite operands to infinite results of the same sign.

**Comment by LilithHafner** [comment]
In Julia master (but not in any released version) we already have

```julia
julia> round(Float32, 1e60)
Inf32

julia> round(Float32, floatmax(Float32)-1.0)
3.4028235f38
```
The method
```julia
round(::Type{T}, x::Real, r::RoundingMode) where {T} = convert(T, round(x, r))
```
was added in #45598 without serious consideration about this behavior. It is not too late to revert or revise this definition.

**Comment by Seelengrab** [comment]
Considering the discussion in https://github.com/JuliaLang/julia/issues/42060, yes that should probably be reverted & limited to `Irrational`.

**Comment by gbaraldi** [comment]
Triage disscussed this and while there are some annoyances to how this works, round documents that it's error throwing behaviour is similar to convert, and if convert doesn't error then neither should round.

**Comment by Seelengrab** [comment]
Was the discussion about how `Irrational` & the corresponding revert is affected?

**Comment by gbaraldi** [comment]
No discussion was made on rounding Irrationals unfortunately. 
What was discussed is that the `round(Float64,1.2)` method is perfectly fine. And that due to how `round` is documented it shouldn't error where convert doesn't. So since convert(Float32,1e60) doesn't error, neither should round, specially because the convert says that if it's representable it will do the conversion and apparently returning `Inf32` means it's representable (That's the IEEE semantic) because we just call a native instruction.

**Comment by nsajko** [comment]
What do you mean by "That's the IEEE semantic"? As far as I understand, IEEE specifies that the result should be chosen among the integers (thus among finite values) except if the input is not finite, as mentioned in my message [above](https://github.com/JuliaLang/julia/pull/50812#issuecomment-1668701537) ("the rounding chooses only from among those floating-point numbers of integral values in the format")?

**Comment by gbaraldi** [comment]
The assembly instructions and the llvm intrinsic all behave like this.

**Comment by Seelengrab** [comment]
That's not the same as IEEE semantic though. As you know, there are a number of operations where LLVM & assembly disagree with IEEE semantic, for example on `min` of non-Nan with Nan values.

**Comment by LilithHafner** [comment]
There should be a test failure in test/rounding.jl due to #51063. We now reuse the mechanics from `convert` (yay!) and `convert` has bugs (not so yay) so we now have bugs (also not so yay).

**Comment by LilithHafner** [comment]
Counting loc in the diff (an imprecise evaluation system) makes me happy: more tests, more docs, less source code, and more functionality :)

**Comment by mgkuhn** [comment]
Could we also have a mode that doesn't throw `InexactError`, and always simply gives us the nearest element of the output type, as in
```
round(UInt8,    +Inf, Nearest) == 0xff
round(UInt8, +1000.0, Nearest) == 0xff
round(UInt8, -1000.0, Nearest) == 0x00
round(UInt8,    -Inf, Nearest) == 0x00
```
where `0xff` and `0x00` are the nearest `UInt8` values to `+Inf`, `1000.0`, `-1000.0` and `-Inf`, respectively. I can think of a lot of applications (signal processing, control systems, graphics), where you never want to have to manually deal with an `InexactError`, and just want to use whatever value is nearest. ([That's how MATLAB converts floats to integers.](https://uk.mathworks.com/help/matlab/matlab_prog/integers.html))

**Comment by LilithHafner** [comment]
That's an interesting idea. This PR is more focused on refactoring fixing holes in the cross product of the existing functionality rather than adding novel functionality, so I split that suggestion off into a separate thread.

**Comment by LilithHafner** [comment]
Just saw #29939. This already fixes half of that issue by accident, but should also fix the other half and add tests.

**Comment by LilithHafner** [comment]
And also just found #51113 when working on fixing https://github.com/JuliaLang/julia/issues/51063. This PR already fixed #51113 by accident when I found it. (optimistically, better wholistic design => less bug surface area => fixes bus I didn't even know about)

Still need to add tests for the issues this fixed by accident.

**Comment by LilithHafner** [comment]
IIUC triage has approved the semantics, but this still needs code review before merging.

As for breaking out the bugfixes for backporting, that is possible, however, 
a) the bugfixes change incorrect answers to errors, which is not something I'm super eager to backport.
b) it's not trivial to separate the bugfixes out and I don't feel personally motivated to do so. None of the bugs were catastrophic.

File: base/float.jl [hunk 1]
```diff
@@ -436,21 +436,25 @@ unsafe_trunc(::Type{UInt128}, x::Float16) = unsafe_trunc(UInt128, Float32(x))
...
-trunc(::Type{Bool}, x::AbstractFloat) = (-1 < x < 2) ? 1 <= x : throw(InexactError(:trunc, Bool, x))
-floor(::Type{Bool}, x::AbstractFloat) = (0 <= x < 2) ? 1 <= x : throw(InexactError(:floor, Bool, x))
-ceil(::Type{Bool}, x::AbstractFloat)  = (-1 < x <= 1) ? 0 < x : throw(InexactError(:ceil, Bool, x))
-round(::Type{Bool}, x::AbstractFloat) = (-0.5 <= x < 1.5) ? 0.5 < x : throw(InexactError(:round, Bool, x))
-
-round(x::IEEEFloat, r::RoundingMode{:ToZero})  = trunc_llvm(x)
-round(x::IEEEFloat, r::RoundingMode{:Down})    = floor_llvm(x)
-round(x::IEEEFloat, r::RoundingMode{:Up})      = ceil_llvm(x)
-round(x::IEEEFloat, r::RoundingMode{:Nearest}) = rint_llvm(x)
+round(::Type{Bool}, x::AbstractFloat, ::RoundingMode{:ToZero}) =
```
Comment (oscardssmith) [inline base/float.jl:445]: It seems like these definitions are valid for all `Integer`s (using typemin/typemax) is there a reson for special casing these?

Comment (LilithHafner) [reply base/float.jl:445]: tehe I underestimated the generic fallbacks. We don't need these methods at all.

Comment (LilithHafner) [reply base/float.jl]: deleted.

**Review by oscardssmith — APPROVED** [review]

**Comment by LilithHafner** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50812#issuecomment-1700274734) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/94a56d2_vs_fb76136/report.html) is available.

**Comment by LilithHafner** [comment]
[DoubleFloats.jl](https://github.com/JuliaMath/DoubleFloats.jl) fails because they don't define `round(::DoubleFloat, RoundDown)`. `round(T, x, RoundDown)` used to be special cased to go straight to `trunc` (unlike other rounding modes wich round first) but with this PR, the generic `round(T, x, RoundDown)` only works if `round(x, RoundDown)` is defined. This is technically a breakage, but idk. I think it's okay because it's due to a [package bug](https://github.com/JuliaMath/DoubleFloats.jl/issues/176) that is literally a typo with a [trivial fix](https://github.com/JuliaMath/DoubleFloats.jl/pull/177).

The other failures look unrelated.

@Seelengrab, I'd rather do t he refactor first and more functionality changes later so that we don't have to keep working with the old system.

**Comment by Seelengrab** [comment]
Why was this PR still marked as closing https://github.com/JuliaLang/julia/pull/50812, when it was clear that this change was not clear cut?

**Pull request merged**